proposals/2024-01-26_classic-histograms-stored-as-native-histograms.md

```diff
@@ -0,0 +1,302 @@
+
+### Scrape
+
+* Create the new representation in a series named `<metric>` without the `le` label. The resulting series name would be the same as if the user enabled exponential histograms. For example `http_request_latency_seconds_bucket`, `http_request_latency_seconds_count` and `http_request_latency_seconds_sum` will simply be stored as a one metric called `http_request_latency_seconds` from now on.
+* Resulting samples at scrape time reuse the existing data structures for native histograms, but use a special schema number to indicate custom buckets. A sample would have either custom bucket definitions or exponential schema and buckets, but not both.
+* If the histogram has exponential buckets during scrape then only the exponential buckets are kept and the custom buckets would be dropped.
+* If all bucket counters and the overall counter of the histogram is determined to be a whole number, use integer histogram, otherwise use float histogram.
+  * In the Prometheus text exposition format values are represented as floats. The dot (`.`) is not mandatory.
+  * In the OpenMetrics text exposition [format](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#numbers) floats and integers are explicitly distinguished by the dot (`.`) sign being present or not in the value.
+  * In the Prometheus/OpenMetrics ProtoBuf [format](https://github.com/prometheus/client_model/blob/d56cd794bca9543da8cc93e95432cd64d5f99635/io/prometheus/client/metrics.proto#L115-L122) float and integer numbers are explicitly transferred in different fields.
+
+### Exemplars
+
+* Reuse the same logic as with exponential histograms. Roughly: parse the exemplars from the classic histogram, discard exemplars without timestamp and sort by timestamp before forwarding.
+
+### Writing/storage
+
+* Remote write protocol to support transferring custom bucket definitions as part of the current Histogram DTO.
+* TSDB should reuse the existing chunk types.
+  * If the sample has exponential schema and buckets, store the exponential schema and buckets as currently.
+  * Otherwise if the sample has custom bucket definition and buckets, use a special schema number and store the custom bucket definitions at the beginning of the chunk once and then the bucket values as before. In the rare case when the bucket layout changes, start a new chunk to only store the custom bucket definitions once per chunk.
+
+### Reading via PromQL
+
+* Direct series selection of `metric{}` would look up `metric{}` and return the exponential histograms or custom histograms as found in the chunks.
+
+  > Note: the [exposition format](https://prometheus.io/docs/prometheus/latest/querying/api/#native-histograms) of histograms in PromQL is already forward compatible with custom buckets as the data contains the boundaries and inclusion settings.
```

Comment:
OTOH the exposition format would contain a log of redundant information especially for range query (matrix) results and in cases where the custom buckets don't change very often.

Maybe a better approach would be to change the exposition format from:
```
"histograms": [ [ <unix_time>, <histogram> ], ... ]

<histogram> = {
  "count": "<count_of_observations>",
  "sum": "<sum_of_observations>",
  "buckets": [ [ <boundary_rule>, "<left_boundary>", "<right_boundary>", "<count_in_bucket>" ], ... ]
}
```
to
```
"histograms": {
  "bounds": {
     0: [ <boundary_rule>, "<left_boundary>", "<right_boundary>"],
     1: [ <boundary_rule>, "<left_boundary>", "<right_boundary>"]
     ...
  },
  "values": [ [<timestamp>, <histogram>], ... ]

<histogram> = {
    "count": "<count_of_observations>",
    "sum": "<sum_of_observations>",
    "buckets": [ ["<boundary_key>", "<count_in_bucket>" ], ... ]
  }
}
```

Reply:
This would also solve an issue with visualization, because without the knowledge about the bucket layout, the UI cannot correctly guess what the possible buckets are. For exponential histograms we should also make sure that 0, 1 bounds are consecutive so that the UI can know the scale, or just explicitly have a key called "scale_factor" , "exponent", whatever.

Reply:
Another idea that's maybe more self contained:
```
<histograms> = [ [ <unix_time>, <histogram> ], ... ]

<histogram> = {
  "count": "<count_of_observations>",
  "sum": "<sum_of_observations>",
  "bounds": [ [ <boundary_rule>, "<left_boundary>", "<right_boundary>"], ... ]
  "counts": [ [<bucket_count>], ...]
}
```

Where "bounds" could be omitted if the "bound" of the sample are equal to the previous "bounds".

This would allow a change in schema/resolution inside one series. It could also still encode scale factor at the end of the "bounds" list, since "counts" would only look up indexes bellow the size of "bounds".

Reply:
I think those are solid ideas, but we probably can postpone it until we the effect of not ideally optimized query/query range API results here. This is also true for exponential buckets, no?

Reply:
Yes, it would be applicable to exponential buckets as well. Probably less effective, but hopefully not worse. I'd have to calculate how many times a schema change needs to happen before this becomes an overhead.

Reply:
From the histograms weekly:  let's postpone this conversation, not part of MVP. Also we're talking about JSON which is verbose anyway, maybe more of an idea for Arrow/ProtoBuf serializations.